src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,655 @@
+
+        _cleanup_free_ char *path_escaped = NULL;
+        _cleanup_free_ char *exec_start = NULL;
+        _cleanup_free_ char *condition_string = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        const char *unit;
+        int r;
+
+        assert(service);
+
+        /* Nothing to do for hidden services. */
+        if (service->hidden) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (service->systemd_skip) {
+                log_info("Not generating service for XDG autostart %s, should be skipped by generator", service->name);
+                return 0;
+        }
+
+        /* Nothing to do if type is not Application. */
+        if (!streq_ptr(service->type, "Application")) {
+                log_info("Not generating service for XDG autostart %s, it is hidden", service->name);
+                return 0;
+        }
+
+        if (!service->exec) {
+                log_info("Not generating service for XDG autostart %s, it is has no Exec= line", service->name);
+                return 0;
+        }
+
+        /*
+         * The TryExec key cannot be checked properly from the systemd unit,
+         * it is trivial to check using find_binary though.
+         */
+        if (service->try_exec) {
+                r = find_binary(service->try_exec, NULL);
+                if (r < 0) {
+                        log_info("Not generating service for XDG autostart %s, TryExec binary %s does not exist or not executable",
+                                 service->name, service->try_exec);
+                        return 0;
+                }
+        }
+
+        r = xdg_autostart_format_exec_start(service->exec_string, &exec_start);
+        if (r < 0) {
+                log_info("Not generating service for XDG autostart %s, Exec line is invalid",
+                         service->name);
+                return 0;
+        }
+
+        if (streq_ptr(service->gnome_autostart_phase, "EarlyInitialization")) {
+                log_info("Not generating service for XDG autostart %s, EarlyInitialization needs to be handled separately",
+                         service->name);
+                 return 0;
+         }
+
+
+        /* Partially handle AutostartCondition (as implemented by GNOME)
+         *
+         * Not implemented are:
+         *  - GSettings (may be updated at runtime)
+         *  - GNOME3 if-session/unless-session (which seems unused and undocumented)
+         */
+        if (service->autostart_condition) {
+                const char *kind;
+                const char *key;
+                const char *state;
+                size_t kind_length;
+                _unused_ size_t key_length;
+
+                /* Split into first word and rest (we don't use key_length). */
+                state = service->autostart_condition;
+                kind = split(&state, &kind_length, WHITESPACE, 0);
```

Comment:
split() is an awful function. Please don#t user that anymore. Please use extract_first_word()